### ISSUE TITLE
Rules counting events using context state are triggered infinitely

### ISSUE THREAD

**Issue by giangiangian** [jruizgit/rules#194]
Hi @jruizgit ,
Thanks for your hard work on this project.
I have been playing with durable_rules for the last week, but I am stuck on a little example I have been playing with.
In particular, I am unable to understand what I am doing wrong, and what would be a feasible way of doing this using durable_rules.

My example situation is pretty simple: I have a week worth of data of eating candies, I can eat at most three candies a day, however, for up to two days a week I can eat up to five candies.
I want to get notified if I ate too many candies.
I tried different approaches, using context states+events and finite machine states.
However, I always end up with either a situation in which my rules are infinitely triggered or are never triggered.

Here is my solution using context states + events:
``` python
from durable.lang import *
import datetime


# I can eat three candies a day every day, except for two days a week when I can eat up to 5
# My rules should notify me of too much candies:
#   - when I ate more than five candies in a single day
#   - when I ate more than three candies in a single day, for the third day in the same week
#
# To simplify, my data covers only a single week (e.g., data goes form monday to next sunday),
# and is time ordered

def get_candies_data():
    def date_string(year, month, day):
        return datetime.datetime(year=year, month=month, day=day).strftime("%d/%m/%y")

    return [
        # I ate:
        # - 4 candies on Monday, <- more than three candies, first day: no notifications
        # - 3 on Tuesday,
        # - 1 on Wednesday,
        # - 6 on Thursday,       <- more than five candies, notification; moreover, second day with more than 3 candies
        # - 4 on Friday,         <- third day with more than three candies: notify me
        # - 2 on Saturday,
        # - 2 on Sunday
        # I expect:
        # - to be notified that on Thursday I ate 6 candies when I could only ate 5
        # - to be notified that on Friday I ate 4 candies when I could only ate 3

        # Monday
        {'verb': 'eat', 'item': 'candy', 'count': 4, 'date': date_string(2019, 2, 25)},
        # Tuesday
        {'verb': 'eat', 'item': 'candy', 'count': 2, 'date': date_string(2019, 2, 26)},
        # Wednesday
        {'verb': 'eat', 'item': 'candy', 'count': 1, 'date': date_string(2019, 2, 27)},
        # Thursday
        {'verb': 'eat', 'item': 'candy', 'count': 6, 'date': date_string(2019, 2, 28)},
        # Friday
        {'verb': 'eat', 'item': 'candy', 'count': 4, 'date': date_string(2019, 3, 1)},
        # Saturday
        {'verb': 'eat', 'item': 'candy', 'count': 2, 'date': date_string(2019, 3, 2)},
        # Sunday
        {'verb': 'eat', 'item': 'candy', 'count': 2, 'date': date_string(2019, 3, 3)},

    ]


with ruleset('candies'):
    # I ate more than 3 candies, but I did not break the rules.
    # Therefore, no notifications here.
    @when_all(s.days_more_than_three_candies < 2,
              c.day << (m.verb == 'eat') & (m.item == 'candy') & (m.count > 3))
    def more_than_three_candies_while_i_could(c):
        # Just update status:
        c.s.days_more_than_three_candies = c.s.days_more_than_three_candies + 1

    # I ate more than 3 candies, and I already did twice this week.
    # Therefore, notify me.
    @when_all(s.days_more_than_three_candies >= 2,
              c.day << (m.verb == 'eat') & (m.item == 'candy') & (m.count > 3))
    def more_than_three_candies_while_i_could(c):
        # Keeping track of further days in which I ate more than three candies is not required,
        # but should not hurt either
        c.s.days_more_than_three_candies = c.s.days_more_than_three_candies + 1

        print("Diet broken! In date {} you could only eat 3 candies, but you ate {}.".format(c.day.date, c.day.count))

    # I ate more than 5 candies, which I can never do.
    # Therefore, notify me. This still counts as a day in which I ate more than 3 candies as well.
    @when_all(c.day << (m.verb == 'eat') & (m.item == 'candy') & (m.count > 5))
    def more_than_three_candies_while_i_could(c):
        # This still counts as a day in which I ate more than three candies
        c.s.days_more_than_three_candies = c.s.days_more_than_three_candies + 1
        print(
            "Diet broken! In date {} you eat more than 5 candies, and in particular you ate {}.".format(c.day.date,
                                                                                                        c.day.count))

    @when_start
    def on_start(host):
        host.patch_state('candies', {'days_more_than_three_candies': 0})
        data = get_candies_data()
        host.post_batch('candies', data)


    @when_all(+s.exception)
    def report_fail(c):
        print('exception {0}'.format(c.s.exception))

run_all()
```
Running this code results in infinite prints of "Diet broken! In date 19/03/01 you could only eat 3 candies, but you ate 4.".

I am aware of #75 , which, if I interpreted correctly, means that each context state change fires a new event.
My idea here was that using events would still have prevented infinite evaluations of the same rule, however, since even if a new state change event is posted, my event of candy eaten is consumed in the process.
However, this seems to not happen. Moreover, even if I remove the state change in said rule, which is not mandatory to solve the problem, the rule is still infinitely called, and I cannot figure out why.

Am I doing something terribly wrong here? Should I be using a different way to solve the problem?
Thank you in advance for your time and help!

**Comment by jruizgit** [comment]
Hi, thanks for posting the question. When I tried your sample code, it did not trigger infinitely. However, it didn't quite give me the results I wanted (due to state changes triggering events). Using facts worked much better for me. 

```python
from durable.lang import *
import datetime


# I can eat three candies a day every day, except for two days a week when I can eat up to 5
# My rules should notify me of too much candies:
#   - when I ate more than five candies in a single day
#   - when I ate more than three candies in a single day, for the third day in the same week
#
# To simplify, my data covers only a single week (e.g., data goes form monday to next sunday),
# and is time ordered

def get_candies_data():
    def date_string(year, month, day):
        return datetime.datetime(year=year, month=month, day=day).strftime("%d/%m/%y")

    return [
        # I ate:
        # - 4 candies on Monday, <- more than three candies, first day: no notifications
        # - 3 on Tuesday,
        # - 1 on Wednesday,
        # - 6 on Thursday,       <- more than five candies, notification; moreover, second day with more than 3 candies
        # - 4 on Friday,         <- third day with more than three candies: notify me
        # - 2 on Saturday,
        # - 2 on Sunday
        # I expect:
        # - to be notified that on Thursday I ate 6 candies when I could only ate 5
        # - to be notified that on Friday I ate 4 candies when I could only ate 3

        # Monday
        {'verb': 'eat', 'item': 'candy', 'count': 4, 'date': date_string(2019, 2, 25)},
        # Tuesday
        {'verb': 'eat', 'item': 'candy', 'count': 2, 'date': date_string(2019, 2, 26)},
        # Wednesday
        {'verb': 'eat', 'item': 'candy', 'count': 1, 'date': date_string(2019, 2, 27)},
        # Thursday
        {'verb': 'eat', 'item': 'candy', 'count': 6, 'date': date_string(2019, 2, 28)},
        # Friday
        {'verb': 'eat', 'item': 'candy', 'count': 4, 'date': date_string(2019, 3, 1)},
        # Saturday
        {'verb': 'eat', 'item': 'candy', 'count': 2, 'date': date_string(2019, 3, 2)},
        # Sunday
        {'verb': 'eat', 'item': 'candy', 'count': 4, 'date': date_string(2019, 3, 3)},

    ]


with ruleset('candies'):
    # I ate more than 3 candies, but I did not break the rules.
    # Therefore, no notifications here.
    @when_all(c.count << m.days_more_than_three_candies < 2,
              c.day << (m.verb == 'eat') & (m.item == 'candy') & (m.count > 3))
    def more_than_three_candies_while_i_could(c):
        # Just update status:
        c.retract_fact(c.count)
        c.assert_fact({'days_more_than_three_candies': c.count.days_more_than_three_candies + 1})
        print('more than three {}, {}'.format(c.s.days_more_than_three_candies, c.day.date))
        
    # I ate more than 3 candies, and I already did twice this week.
    # Therefore, notify me.
    @when_all(c.count << m.days_more_than_three_candies >= 2,
              c.day << (m.verb == 'eat') & (m.item == 'candy') & (m.count > 3))
    def more_than_three_candies_while_i_could(c):
        # Keeping track of further days in which I ate more than three candies is not required,
        # but should not hurt either
        c.retract_fact(c.count)
        c.assert_fact({'days_more_than_three_candies': c.count.days_more_than_three_candies + 1})
        print("Diet broken! In date {} you could only eat 3 candies, but you ate {}.".format(c.day.date, c.day.count))

    # I ate more than 5 candies, which I can never do.
    # Therefore, notify me. This still counts as a day in which I ate more than 3 candies as well.
    @when_all(c.count << +m.days_more_than_three_candies,
              c.day << (m.verb == 'eat') & (m.item == 'candy') & (m.count > 5))
    def more_than_three_candies_while_i_could(c):
        # This still counts as a day in which I ate more than three candies
        c.retract_fact(c.count)
        c.assert_fact({'days_more_than_three_candies': c.count.days_more_than_three_candies + 1})
        print(
            "Diet broken! In date {} you eat more than 5 candies, and in particular you ate {}.".format(c.day.date,
                                                                                                        c.day.count))

    @when_start
    def on_start(host):
        host.assert_fact('candies', {'days_more_than_three_candies': 0})
        data = get_candies_data()
        host.post_batch('candies', data)


    @when_all(+s.exception)
    def report_fail(c):
        print('exception {0}'.format(c.s.exception))

run_all()
```

**Comment by giangiangian** [comment]
Hi @jruizgit , thank you for your answer! Facts work much better for me as well, however I still had the problem of infinite prints on your version as well.

After some headaches, I was able to fix it by cleaning my Redis setup (command `redis-cli FLUSHALL`, in particular); running your code now gives me the expected results.
I think I did some horrible mistakes at some point, while learning how to use rules, and created a huge queue of facts/events.  Hope this clue may help others with same symptoms.

Thanks again! I'll close this issue as I do not have an issue anymore :smile:
